File: backend/app/schemas/rag.py
from typing import Literal

from pydantic import BaseModel, Field

GroundingPolicy = Literal[
    "structured_claim_refs_v3", "anthropic_tool_use_v1", "opencode_go_tool_use_v1"
]


class CitationCard(BaseModel):
    literature_id: str
    chunk_id: str | None = None
    title: str
    source: str
    snippet: str
    quote: str | None = None
    reason: str | None = None
    confidence: float
    match_score: float | None = None
    source_type: str | None = None
    pdf_upload_id: str | None = None
    related_entity_ids: list[str] = Field(default_factory=list)


class RetrievalMetadata(BaseModel):
    applied_source: Literal["all", "cn_literature", "pubmed"]
    applied_top_k: int
    available_citation_count: int
    strategy: str = "keyword"


class GroundedClaim(BaseModel):
    text: str
    evidence_refs: list[str] = Field(default_factory=list)
    semantic_score: float | None = None
    entailment_score: float | None = None


class GroundingMetadata(BaseModel):
    status: Literal["skipped", "passed", "blocked"]
    policy: GroundingPolicy = "structured_claim_refs_v3"
    checked: bool
    blocked_reason: str | None = None
    allowed_evidence_refs: list[str] = Field(default_factory=list)
    matched_evidence_refs: list[str] = Field(default_factory=list)
    unsupported_evidence_refs: list[str] = Field(default_factory=list)
    claim_count: int = 0
    cited_claim_count: int = 0
    structured_claims: list[GroundedClaim] = Field(default_factory=list)
    provider_native_grounding: bool = False
    tool_name: str | None = None
    tool_call_count: int = 0
    semantic_threshold: float | None = None
    min_semantic_score: float | None = None
    nli_threshold: float | None = None
    min_entailment_score: float | None = None


class RagAnswerRequest(BaseModel):
    question: str = Field(min_length=1)
    source: Literal["all", "cn_literature", "pubmed"] = "all"
    top_k: int = Field(default=2, ge=1)


class ProviderSli(BaseModel):
    """Service-level indicators for the LLM provider call.

    ``provider_latency_ms`` wraps only the ``generate_answer`` call (retrieval and
    grounding are local and cheap). ``estimated_cost_usd`` is derived from token
    usage and operator-configured per-million-token prices; it stays ``None`` when
    token usage is missing (deterministic / fallback) or prices are unset, so we
    never surface a guessed price.
    """

    provider_latency_ms: int | None = None
    estimated_cost_usd: float | None = None


class RagAnswerResponse(BaseModel):
    question: str
    answer: str
    disclaimer: str
    retrieval: RetrievalMetadata
    citations: list[CitationCard]
    answered_at: str
    provider_name: str
    grounding: GroundingMetadata
    input_tokens: int | None = None
    output_tokens: int | None = None
    sli: ProviderSli | None = None
    integrity_token: str | None = None

File: frontend/lib/api/rag.ts
export type RagSource = "all" | "cn_literature" | "pubmed";

export type CitationCard = {
  literature_id: string;
  chunk_id?: string | null;
  title: string;
  source: string;
  snippet: string;
  quote?: string | null;
  reason?: string | null;
  confidence: number;
  source_type?: string | null;
  pdf_upload_id?: string | null;
  related_entity_ids?: string[];
};

export type RetrievalMetadata = {
  applied_source: RagSource;
  applied_top_k: number;
  available_citation_count: number;
  strategy: string;
};

export type GroundedClaim = {
  text: string;
  evidence_refs: string[];
  semantic_score?: number | null;
};

export type GroundingMetadata = {
  status: "skipped" | "passed" | "blocked";
  policy: "structured_claim_refs_v3" | "anthropic_tool_use_v1" | "opencode_go_tool_use_v1";
  checked: boolean;
  blocked_reason?: string | null;
  allowed_evidence_refs: string[];
  matched_evidence_refs: string[];
  unsupported_evidence_refs: string[];
  claim_count: number;
  cited_claim_count: number;
  structured_claims: GroundedClaim[];
  provider_native_grounding: boolean;
  tool_name?: string | null;
  tool_call_count: number;
  semantic_threshold?: number | null;
  min_semantic_score?: number | null;
};

export type ProviderSli = {
  provider_latency_ms?: number | null;
  estimated_cost_usd?: number | null;
};

export type RagAnswerResponse = {
  question: string;
  answer: string;
  disclaimer: string;
  retrieval: RetrievalMetadata;
  citations: CitationCard[];
  answered_at: string;
  provider_name: string;
  grounding: GroundingMetadata;
  input_tokens?: number | null;
  output_tokens?: number | null;
  sli?: ProviderSli | null;
};

export function getBackendBaseUrl() {
  return process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://127.0.0.1:8000";
}

export function buildRagAnswerUrl() {
  return new URL("/api/rag/answer", getBackendBaseUrl()).toString();
}

export function buildRagAnswerRequest(question: string, source: RagSource = "all", topK = 2) {
  return {
    question: question.trim(),
    source,
    top_k: topK,
  };
}

export function getRagSourceLabel(source: RagSource) {
  if (source === "cn_literature") {
    return "中文文献";
  }
  if (source === "pubmed") {
    return "PubMed";
  }
  return "全部文献";
}

export async function answerRagQuestion(
  question: string,
  source: RagSource = "all",
  topK = 2,
): Promise<RagAnswerResponse> {
  const response = await fetch(buildRagAnswerUrl(), {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(buildRagAnswerRequest(question, source, topK)),
  });

  if (!response.ok) {
    throw new Error("RAG answer request failed");
  }

  return response.json();
}
